### PR TITLE
feat: Adds `beforeUnload` event handler to survey editor so that users don't accidentally lose their survey edit progress if they refresh or navigate away

### LINKED ISSUE #22286: Alert me on refresh so that I don’t lose all of my progress when refreshing the page
## Is your feature request related to a problem?

Right now, refreshing the survey editor form kills all of your progress.  This is distressing for users who've spent a good chunk of time working on it (e.g. editing survey rules).  We should have some sort of alert that prevents/warns users from refreshing the page so they're aware that doing so nukes their progress.

## Describe the solution you'd like

@jurajmajerik mentioned some other parts of the product have an existing UX for this, so I would propose just piggybacking on that.

### Debug info

```shell
- [ ] PostHog Cloud, Debug information: [please copy/paste from https://us.posthog.com/settings/project-details#variables]
- [ ] PostHog Hobby self-hosted with `docker compose`, version/commit: [please provide]
- [ ] PostHog self-hosted with Kubernetes (deprecated, see [`Sunsetting Kubernetes support`](https://posthog.com/blog/sunsetting-helm-support-posthog)), version/commit: [please provide]
```

### PR DESCRIPTION
**Pull request by dmarticus**
This change does the following

1. adds a `beforeUnload` handler to the survey editing experience so that users don't accidentally refresh the page and lose their survey edit progress.  As part of implementing this, I fixed an issue where the initial survey creation process never set the isEditingSurvey state to true.
2. fixes the behavior for `cancel` on new surveys; before these changes, clicking `cancel` on a new survey would take you to this weird ephemeral state of this new survey.  Now, clicking cancel on a new survey that hasn't been saved just takes you back to the `/surveys` URL.  [See this message for context](https://posthog.slack.com/archives/C034XD440RK/p1716327059331599).
3. Removes the the `Cancel` and `Save` buttons from the footer.  Felt unnecessary.  See [this comment](https://github.com/PostHog/posthog/pull/22375#discussion_r1607438057) for details.

Overall, this closes https://github.com/PostHog/posthog/issues/22286

## Changes

Here's the new survey editing experience, with these changes applied

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

I didn't write any new unit/e2e tests, and it's totally possible that I should.  Might sync with the team on how to do those.  Here were the test cases I ran, for posterity:

### FAILING TESTS

- The reason this PR hasn't been merged yet is because the following scenario is failing
- User navigates to `/surveys` -> clicks on an existing survey -> goes to edit that existing survey -> hits the "back" button on the browser -> beforeunload starts looping

#### New surveys
- Creating a new survey with the desired flow doesn’t trigger any unwanted modals
- Creating a new survey and then cancelling the creation takes you back to the surveys page
- Creating a new survey, making some edits to it, and then trying to leave the page throws the alert
- Creating a new survey, making some edits to it, and then saving it as a draft does not trigger the alert, and navigates you to the survey overview for that new survey

#### Existing surveys
- Cancelling an edit session for an existing survey takes you to that survey overview
- Attempting to leave an edit session for an existing survey throws the modal
- Hitting “save” on edits for an existing survey doesn’t throw the modal, and instead takes you to the overview of that survey.
- Hitting “cancel” on edits for an existing survey takes you to the survey overview without the edits applied.

### CODE REVIEW

File: frontend/src/scenes/surveys/Survey.tsx [hunk 1]
```diff
@@ -28,7 +29,13 @@ export const scene: SceneExport = {
...
 export function SurveyComponent({ id }: { id?: string } = {}): JSX.Element {
     const { isEditingSurvey, surveyMissing } = useValues(surveyLogic)
-    const showSurveyForm = id === 'new' || isEditingSurvey
+    const { editingSurvey } = useActions(surveyLogic)
+
+    useEffect(() => {
+        if (id === 'new') {
+            editingSurvey(true)
+        }
+    }, [id, editingSurvey])
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/Survey.tsx:38]: I reached for `useEffect()` here since that's my classic tool for this type of react state management.  If there's a more posthoggy way to do this, plz let me know.

File: frontend/src/scenes/surveys/Survey.tsx [hunk 2]
```diff
@@ -88,26 +95,6 @@ export function SurveyForm({ id }: { id: string }): JSX.Element {
...
-                            router.actions.push(urls.surveys())
-                        }
-                    }}
-                >
-                    Cancel
-                </LemonButton>
-                <LemonButton type="primary" data-attr="save-survey" htmlType="submit" loading={surveyLoading}>
-                    {id === 'new' ? 'Save as draft' : 'Save'}
-                </LemonButton>
-            </div>
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/Survey.tsx:110]: Drive-by change that's not really in scope, but Boy Scout Principle and all that. I noticed that we had a footer on this page, which felt superfluous given that the page has a sticky header.  Here's what it looks liek with the footer

Screenshot 2024-05-20 at 5 07 21 PM

And what it looks like without

Screenshot 2024-05-20 at 5 06 59 PM

I prefer without.

File: frontend/src/scenes/surveys/surveyLogic.tsx [hunk 3]
```diff
@@ -785,6 +785,12 @@ export const surveyLogic = kea<surveyLogicType>([
             return [urls.survey(values.survey.id), router.values.searchParams, hashParams]
         },
     })),
+    beforeUnload(({ values }) => ({
+        enabled: (newLocation) => {
+            return values.isEditingSurvey && newLocation?.pathname !== router.values.location.pathname
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyLogic.tsx:790]: the router stuff probably isn't super necessary; really, all we care about if if the user is in an edit session or not.  I just followed the pattern from some of the other `beforeUnload` handlers in our different scenes.

File: frontend/src/scenes/surveys/Survey.tsx [hunk 4]
```diff
@@ -51,6 +58,19 @@ export function SurveyForm({ id }: { id: string }): JSX.Element {
...
+            if (id === 'new') {
+                router.actions.push(urls.surveys())
+            } else {
+                loadSurvey()
+            }
+        } else {
+            router.actions.push(urls.surveys())
+        }
+    }
+
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/Survey.tsx:73]: This fixes the `Cancel` button issue I called out here: https://posthog.slack.com/archives/C034XD440RK/p1716327059331599

File: frontend/src/scenes/surveys/surveyLogic.tsx [hunk 5]
```diff
@@ -785,6 +790,17 @@ export const surveyLogic = kea<surveyLogicType>([
             return [urls.survey(values.survey.id), router.values.searchParams, hashParams]
         },
     })),
+    beforeUnload(({ values, actions }) => ({
+        enabled: (newLocation) => {
+            return values.isEditingSurvey && newLocation?.pathname !== router.values.location.pathname
+        },
+        message:
+            'You have unsaved changes to your survey. If you leave this page, your changes will not be saved. Are you sure you want to proceed?',
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyLogic.tsx:795]: This message seems reasonable enough to me, but maybe too wordy?

**Review by Phanatic — APPROVED** [review]
Looks good to me, great work!

File: frontend/src/scenes/surveys/Survey.tsx (hunk 1 above)
Comment (dmarticus) [reply frontend/src/scenes/surveys/Survey.tsx:38]: @neilkakkar I might ask for your advice on this.  I know our [coding conventions](https://posthog.com/handbook/engineering/conventions/frontend-coding) say to not use `useEffect` to store state, but in my case here, I can't figure out where to insert the editing process into the survey flow (this is just for new surveys, fwiw).  Let me provide some context.

Right now, the way survey editing works is that whenever we click "New survey" and are prompted with the new survey creation page, we have the `id` prop set to `'new'`, but we haven't entered the survey editing experience yet.  When I tried removing this `useEffect()` and replacing it by just calling the action inside the `Create Blank Survey` button
```
            <PageHeader
                buttons={
                    <LemonButton
                        type="primary"
                        onClick={() => editingSurvey(true)}
                        to={urls.survey('new')}
                        data-attr="new-blank-survey"
                    >
                        Create blank survey
                    </LemonButton>
                }
            />
```

I kept accidentally triggering the `beforeunload` handler, since the editing mode was applied before the new URL loaded.  

Given how surveys are designed now, what I'd need to do is have some sort of way to set editing state _after_ we've landed on the new survey URL.  I'm not sure how to best do that using Kea stuff, so I'd love your input on how to do this the PostHog way.

Right now, what I'm doing works, but it definitely does break our survey editing state management abstraction, and I'm not 100% certain what the appropriate way of resolving this is going forward.

File: frontend/src/scenes/surveys/surveyLogic.tsx (hunk 3 above)
Comment (mariusandra) [reply frontend/src/scenes/surveys/surveyLogic.tsx:792]: Flyby: Beforeunloads are notoriously flakey. We've been getting support requests for the one on insights ever since we introduced it. We haven't done the work yet, but the plan is to move to a "store the unsaved state in the URL" approach. This way if you accidentally click away from a saved state, you can click back and then save it. Using actionToUrl with `{replace: true}` as the last parameter should make this kind of easy/possible.

Obviously feel free to ignore me :).

File: frontend/src/scenes/surveys/Survey.tsx (hunk 1 above)
Comment (neilkakkar) [reply frontend/src/scenes/surveys/Survey.tsx:38]: Look into `urlToAction` in the surveylogic! I think adding this to the end there should work - it effectively means when you end up on URL X, fire action Y.

And makes sense, on the button would be too soon indeed 👍

File: frontend/src/scenes/surveys/surveyLogic.tsx (hunk 3 above)
Comment (dmarticus) [reply frontend/src/scenes/surveys/surveyLogic.tsx:792]: I appreciate the feedback!  Neil had a similar comment about this in my implementation, so I'm going to have a whack at trying that approach instead.

File: frontend/src/scenes/surveys/Survey.tsx (hunk 1 above)
Comment (dmarticus) [reply frontend/src/scenes/surveys/Survey.tsx:38]: oh, that's very cool, and sounds like exactly what we're looking for.  Let me give that a shot!

Comment (dmarticus) [reply frontend/src/scenes/surveys/Survey.tsx:38]: Okay, so that _almost_ worked, I made this change to the `urlToAction` hook
```
    urlToAction(({ actions, props }) => ({
        [urls.survey(props.id ?? 'new')]: (_, __, _e__, { method }) => {
            // If the URL was pushed (user clicked on a link), reset the scene's data.
            // This avoids resetting form fields if you click back/forward.
            if (method === 'PUSH') {
                if (props.id) {
                    actions.loadSurvey()
					// new stuff
                    if (props.id === 'new') {
                        actions.editingSurvey(true)
                    }
                } else {
                    actions.resetSurvey()
                }
            }
        },
    })),
```

And it worked for 7/8 of my test cases.  However, it's getting stuck in this loop where if I navigate to an existing survey from the `/surveys` URL, enter editing mode, and then try to hit the "back" button, I'm stuck in an infinite beforeunload loop because it keeps toggling between the 2 URLs (see attached gif).  I'm not sure how to get around it.

At this point, I'm still kinda included to go with `useEffect()` on the survey page because it always works, and then clean up this state management as I keep going.  The idea here is that I fix the broken behavior immediately and then fix the state abstractions once the user pain is mitigated.  What do you think?

Comment (neilkakkar) [reply frontend/src/scenes/surveys/Survey.tsx:38]: d'oh, annoying, yeah agree sounds good, let's go with the useEffect for now, I'll take another look at this on Monday too - unclear why it would get stuck 😓

Comment (dmarticus) [reply frontend/src/scenes/surveys/Survey.tsx:38]: Oh very lame, this is happening with `useEffect()` too, actually (gif example with the URLs included so you can see what I mean).  The specific use case is when a user goes from surveys page -> clicks on existing survey -> clicks "edit" -> tries specifically to click the "back" button in the browser (which sends them from the `/surveys/{id}` URL to `/surveys`.  

Gonna get back in the lab for this, I don't think this behavior is an acceptable fix.

Comment (neilkakkar) [reply frontend/src/scenes/surveys/Survey.tsx:38]: spent some time looking into this, this is too flakey indeed, and I couldn't figure this out. I'm leaning a lot more towards Marius's suggestion now indeed - we can base64 encode the entire payload into the URL, so when you click back, and then forward, the url populates the data again, seems much cleaner 👍
